### ISSUE TITLE
Support Rescanning Files / Adding & Removing File Nodes from the Graph

### ISSUE THREAD

**Issue by avose** [avose/GLShell#17]
Right now, files are scanned once at application launch and are then static. We need a "rescan" option to look for new files or removed files.

**Comment by avose** [comment]
This has a pretty good start, but there are some issues. Performance is one. I need to rework the way some of this is done to get reasonable performance out of it. It may not matter too much, as there is a limit imposed by the speed of the physics calculation.. but lots of rapid changes in the watched directory is going to cause lots of repeated scanning, and this can be slow.

There is also a possible index out of bounds error where the graph rendering is indexing into a node from an edge, and the index is out of bounds. Things work like.. pretty much all the time.. but I have seen a failure or two. I was thinking of thread synchronization issues, but I think I'm holding the lock when drawing and also holding the lock when making any updates to the graph, so I'm surprised to see this. Maybe something else is wrong.

Really, since there are performance issues anyway, maybe this whole thing needs to be looked over again and potentially have large parts of it re-written. When it was first done, I wasn't really thinking of making changes to things after they had been read, so.. here we are.

There are also some issues when rapid changes are being made, as I'll do a "watcher.AddTree()" to add a newly created directory, but by the time I get around to adding it, the directory could have been removed. This should at least be caught with some "try / except" handling so as not to crash things. A rescan can just be scheduled for later in hopes things settle down.

Actually... I only rescan the filesystem to look for new files / dirs... and I'm relying on the deletion events from the watcher to see when things are removed. Since the watcher isn't reliable, I need to handle removal myself when things are rescanned.

In short, this whole process needs a lot of work. I should think about this a while before I mess with it, and then probably rewrite the whole thing...

**Comment by avose** [comment]
This has been cleaned up a little bit. I haven't seen any crashes again with more testing than I have ever done before. It's still not perfect, and it could still end up being very slow.

However, I think the speed won't matter much, because the physics calculation will be the limiting factor, and there's shouldn't be a whole bunch of rapid churn all the time in a watched directory -- the idea is to use this as a source code / project editor, so besides some file copies and builds, and some "make clean"s, etc., it should be OK.

I'm going to go ahead and close this one, as for normal usage, it seems to be working fine.

I'll open a new issue about code cleanup / performance, maybe I get to it again at a later date.
